### DISCUSSION TITLE
Questions for using ERA-5 atmos forcing for global HRLDAS/Noah-MP simulations

### DISCUSSION THREAD

**Discussion by cenlinhe** [NCAR/hrldas#87] [Q&A]
The following is copied from communications between a user and our Noah-MP technical team regarding using ERA-5 atmos forcing for global HRLDAS/Noah-MP simulations, which may be useful for other users in the community.

---
Q: "I am using the script extract_era5.py to create the forcings, but all the output files that are created have 0 size. I think there is an issue in the dforecast and danalysis strings used to make the grep."

A: "The 0 size files mean the selected date or variable names are not included in the forcing file. Looks like you downloaded the single-level ERA5 (https://cds.climate.copernicus.eu/cdsapp#!/dataset/reanalysis-era5-single-levels?tab=overview) but not ERA5-LAND (https://cds.climate.copernicus.eu/cdsapp#!/dataset/reanalysis-era5-land?tab=form). However, our current HRLDAS_forcing ERA5 scripts only support the use of ERA5-Land. You could also see this issue: https://github.com/NCAR/hrldas/issues/56
Even for the same variable, 10U, the variable name/identifier discrepancies in the two files (ERA5-atmos vs. ERA5-land) cause the "0 size" issue when running the code. So it turns out, different ways of getting data will cause problems when extracting variables. 
Note that there are a lot of differences between single-level ERA5 and ERA5-Land products, e.g. variable name, lat, lon resolution, etc. 
For now, our scripts in GitHub only work with ERA5-land, as it has higher spatial resolution (0.1degree)."

---
Q: "I downloaded the new forcing and now the python scipts work fine. However, not I am stuck with create_forcing.exe: in particular, I get the following error message:

CHECK_IF_SAME_MAP:
     proj_code=            0           0
     nx       =         3600        3600
     ny       =         1801        1801
     startlat =      90.000000000000     90.000000000000
     startlon =     179.999984741211      0.000000000000
     deltalat =      -0.100000001490     -0.100000001490
     deltalon =       0.100000001490      0.100000001490
     dx       =    -999.900024414062   -999.900024414062
     dy       =    -999.900024414062   -999.900024414062
     xlonc    =    -999.900024414062   -999.900024414062
     truelat1 =    -999.900024414062   -999.900024414062
     truelat2 =    -999.900024414062   -999.900024414062
1) Source model terrain dimensions do not match source model temperature dimensions
I also applied this fix to the code, but it doesn't fix the issue. 
https://github.com/NCAR/hrldas/pull/78
Any other fix for this? Then, if a play with input files modifying the coordinates I get this error message:

RAINRATE   GRIB Editon 1
grib%g1sec2%j_scan_direction =           -1
             Maxval adjusted from     0.010064 to     0.002796:    factor of        4
             :  Checking for file 'ERA5/OUTPUTS/TP/ERA5_TP_1980010100'
             :  Found file ERA5/OUTPUTS/TP/ERA5_TP_1980010100
RAINRATE   GRIB Editon 1
grib%g1sec2%j_scan_direction =           -1
             Maxval adjusted from     0.010064 to     0.002796:    factor of        4
 You have this many undefined T points over valid land points:       966487
 This tends to happen around coastlines or for unresolved islands in your forcin
 g data
 ===== Consider increasing expand_loop in namelist =====
 However, this may also occur if your forcing source does not cover your domain.
I wonder what I am doing wrong and how can I solve these issues.
"

A: "This error seems to be caused by the inconsistent ERA-5 terrain data used with ERA5-Land. I assume you are creating a global domain, right? You may also want to check if the lat/lon max and min defined in ERA-5 is wider than your global domain lat/lon defined so that the forcing map fully covers your study domain."

Q: "although I am using global ERA5-Land data, the error in check_if_same_map is caused by a shift in longitude between the ERA5-Land geopotential and the ERA5-Land T2 (please see the attached figures).
For this reason, you developed a bug fix at L1262 of [hrldas/hrldas/HRLDAS_forcing/create_forcing.F.](https://github.com/NCAR/hrldas/blob/20be331a32f618f15a3c4b542608e03190be9661/hrldas/HRLDAS_forcing/create_forcing.F#L1262)
However, to my understanding, the call to the subroutine remap is below L958, namely where I get the error message "stop "1) Source model terrain dimensions do not match source model temperature dimensions"". So, at the call to check_if_same_map, the code still finds z and T2 on different grids and it stops working. Does it make sense, or am I missing something?"

A: "Yes, you are right. The geopotential is read before remapping, and the geopotential longitudes also need to be the same as temperature grib file between 0 and 360 degrees. In addition, the geopotential file has to be the NetCDF format.  The bug fix is due to the longitude from the WPS geofile map being between -180 and 180. So when remapping to the regional domain for geogrid, I changed the longitude to plus 360 degree if the longitude is less than zero from the geogrid file.
if your T2 longitude grid is between -180 and 180. You will have to change the geo and lsm file to -180 and 180 as well and the grid info in subroutine read_era5_netcdf_unit
Also,  the +360 is not required at L1262 of [hrldas/hrldas/HRLDAS_forcing/create_forcing.F.](https://github.com/NCAR/hrldas/blob/20be331a32f618f15a3c4b542608e03190be9661/hrldas/HRLDAS_forcing/create_forcing.F#L1262) as the geogrid longitude is also between -180 and 180."

---
Q: "These changes allow to solve the issue, but now a new problem is popping up:

 You have this many undefined T points over valid land points:       966487
 This tends to happen around coastlines or for unresolved islands in your forcin
 g data
 ===== Consider increasing expand_loop in namelist =====
 However, this may also occur if your forcing source does not cover your domain.

I tried also to enable the MASK_WATER = .TRUE. but it didn't solve the problem. That's odd as my Noah-MP domain (attached) is smaller than ERA5, which is global."

A: "Can you try changing the following three things each time.
1. change read_era5_netcdf_unit, 
lon1=180.00 to lon1=-180
2. remove the +360. at L1262 of [hrldas/hrldas/HRLDAS_forcing/create_forcing.F](https://github.com/NCAR/hrldas/blob/20be331a32f618f15a3c4b542608e03190be9661/hrldas/HRLDAS_forcing/create_forcing.F#L1262)
if  in the latest version,  remove the + 360. at L1263, https://github.com/NCAR/hrldas/blob/1c695d60c69ec44513c523a5fe4ab148ee24fe80/hrldas/HRLDAS_forcing/create_forcing.F#L1263
3. do both 1 and 2 at the same time

Your latest error message indicates that some of the coastal pixels or small islands are not fully resolved during spatial interpolation in the forcing pre-processing, so please increase the "expand_loop" entry value in your namelist. You can use expand_loop = 30 or 50 if it does not solve the issue. As you increase this number, it will take relatively more time to process all the forcing, but it will be more accurate in terms of spatial interpolation.

I only change the EXPAND_LOOP in the namelist, when increasing the EXPAND_LOOP, the missing grid point is reduced. These grid points are somewhere where the forcing data is ocean but geogrid is land. I have tried to increase EXPAND_LOOP to 260, the error is finally gone, and forcing data can be generated over time.
"

---
Q: "playing a bit with data (I regridded ERA5-Land meteo forcings to the geopotential grid) and restoring the original code (i.e no changes in read_era5_netcdf_unit) I have been able to overcome the issue; however, I had to increase EXPAND_LOOP=360 and the program became really slow. Considering that I have to prepare more than 30-year data, I wonder if there is a way to improve this task, e.g. may I change the namelist.wps to create the domain in a different way or other solutions?"

A: "Some of way I can think of is to make the Same land mask of Georgia file as ERA5 land mask, so there is no need to interpolate ERA5–Land data, and can remove this part of code in the Fortran
Or just not to do the EXPAND_LOOP and remove this error in the Fortran, but do the post processing using other ways for interpolation, like cdo or python.
I think if you have the same geogrid landmask with ERA5-Land forcing landmask, then you do not need to remove the interpolation code but just set  EXPAND_LOOP to 0 and it should work."

---
Q: "I think specifying the area in this way into the python script leads to the issue. 
finally, downloading the ERA5-Land data without specifying the area, I have been able to replicate your results (i.e. with EXPAND_LOOP = 260 I have been able so solve the issue) and no modifications to the code are needed. However, now I am getting this error:

Field label= SNOW
No previous data.

I tried to chance the dates, but it won't work. Did you get this message?

As for the domain, I am running geogrid with this configuration:

&geogrid
 parent_id         = 1,
 parent_grid_ratio = 1,
 i_parent_start    = 1,
 j_parent_start    = 1,
 e_we              = 3601,
 e_sn              = 1802,
 geog_data_res     = 'default',
 dx                = 0.1,
 dy                = 0.1,
 map_proj          = 'lat-lon',
 stand_lon         = 0.
 pole_lat          = 90.0
 pole_lon          = 0.0
 ref_lat           = 0.0,
 ref_lon           = 179.95,

Considering the longitude, this set-up produces a domain similar to ERA5, exept for a few digits (see attached file); however, the ERA5 latitude is:
90 to -90 by -0.1 degrees_north  

while with geogrid I get the opposite:
-90 to 90 by 0.1 degrees_north  

So, at some point I need to flip one of the two to avoid interpolation; is it right?
"

A: "The error is due to generating the initialization file, if the FULL_IC_FRQ in the namelist is set to 0 or larger. 
0 is to create full initial conditions at the starting time only
If this is set to -1, the error will be gone, and not using ERA5-Land for initialization.

Please see /hrldas/docs/README.ERA5 for more information. Here I copy from there. 

To use ERA5-Land data for initialization, please also download extra variables, but only for the first timestep:
Skin temperature, Soil temperature (level1-4), Snow depth water equivalent, Volumetric soil water (layer1-4)
(ERA5-Land doesn't provide Canopy water storage for initialization)
extract_era5_init.py: extract the initial-only fields (SWE_inst,Canopint_inst,AvgSurfT_inst,SoilMoi,SoilTMP) from the ERA5 output files
Change the date of initialization and run the script one time. The files should appear in the /extracted_ERA5_data_path/INIT/ directory.
Note: this step is not necessary unless you want to initialize using the ERA5 states
Run the python code:
`python extract_era5_init.py`
"

---
Q: "***DATE=1980-01-01_00:00:00 Write restart at 1980-01-01_00

NetCDF: One or more variable sizes violate format constraints                  

 ***** Problem nf90_enddef

Stopped

Any idea on what is going wrong?"

A: "I found the issue with netcdf creation: I realized that -DWRFIO_NCD_LARGE_FILE_SUPPORT was not specified into user_build_options, therefore in prepare_output_file_seq
the output file was created without the large file support (i.e. iret = nf90_create(trim(output_flnm), NF90_CLOBBER, ncid)) and, being the output pretty large, I got the error at the end of its definition (i.e nf90_enddef).
I put the -DWRFIO_NCD_LARGE_FILE_SUPPORT and compiled again and the model is running fine."

**Comment by cenlinhe** [comment]
The filename of the extracted grib files for each variable have an hour early shift compared to the actual time in the file.
see below screenshot, e.g., ERA_10U_2020010123 and ERA_2T_2020010200. We should also discuss this later.

**Comment by tslin2** [comment]
Q: "The problem I have is related to SoilWatConductivity when SURFACE_RUNOFF_OPTION=1 ( I only debugged this option, but similar errors occur when using other runoff schemes different from 3). When I use this configuration the model crashes at the first time step; I found out that SoilWatConductivity=0, thus the model stops with a segmentation fault into GroundWaterTopModelMod.f90. Looking at SoilHydraulicPropertyMod.F90, the problem seems originating in this equation when the SoilMoisture is near zero:

SoilPreFac = max(0.01, SoilMoisture/SoilMoistureSat(IndLayer))

These are the values of the variables used in the equation before the models stops: print*,'SoilPreFac',SoilPreFac,SoilMoisture,SoilMoistureSat(IndLayer),SoilMoisture/SoilMoistureSat(IndLayer)

SoilPreFac  1.1815587E-02  5.1870430E-03  0.4390000      1.1815587E-02

Being SoilPreFac so small, I get in the following equation SoilWatConductivity=3.2141289E-32, which for round off becomes 0:   
SoilWatConductivity = SoilWatConductivitySat(IndLayer) * SoilPreFac ** SoilExpTmp

Any idea why this happens?
"

A: "I agree with you that this is a precision issue. In the GroundWaterTopModelMod.F90, there is division by WatConductAcc (sum of SoilWatConductTmp*ThicknessSoil). Because WatConductAcc is zero, the model reports floating invalid error.
I did some tests in my case at 0.1 degree and found out that it happened around central Australia with SILTY CLAY LOAM soil in barren land, where soil moisture is extremely low.
I tried to change the 0.01 to 0.06 in this equation, SoilPreFac = max(0.01, SoilMoisture/SoilMoistureSat(IndLayer)). In this revision, the model runs fine with single precision and does not crash.

I have been able to solve it replacing the compilation flag -fpe0 with -fp-model=strict.  Anyway, as this error takes place under near zero soil moisture conditions, I guess the above suggestion is the optimal solution.
"
